### PR TITLE
Feature: Make slack request rate configurable

### PR DESCRIPTION
**Pull request by ani1311**
If multiple pods of this app is deployed, they will not respect the slack API request rate of 1 RPS. To have multiple pods of this app using the same slack bot token, we need to have the SlackRequestRPS configurable.

### CODE REVIEW

**Comment by wiardvanrij** [comment]
- [ ] Rebase
- [ ] Please fix unit tests and if applicable, add more
- [ ] Please update the readme, so it includes this new config option

Thanks :D

File: app.go [hunk 1]
```diff
@@ -162,11 +162,11 @@ func (app *App) Shutdown() {
 }
 
 //nolint:gocognit // but could probably use a refactor.
-func (app *App) processQueue(ctx context.Context, maxRetries int, initialBackoffMs int, slackPostMessageURL string, tokenFlag string, burst int) {
+func (app *App) processQueue(ctx context.Context, maxRetries int, initialBackoffMs int, slackPostMessageURL string, tokenFlag string, burst int, slackRequestRateMs int) {
```
Comment (ldemailly) [inline app.go:165]: This starts to be too many parameters and could be a Config object (with the extra benefit that you could use struct2env to get values from env for it automatically)

File: app.go [hunk 2]
```diff
@@ -162,11 +162,11 @@ func (app *App) Shutdown() {
...
 	// This is the rate limiter, which will block until it is allowed to continue on r.Wait(ctx).
 	// I kept the rate at 1 per second, as doing more than that will cause Slack to reject the messages anyways. We can burst however.
 	// Do note that this is best effort, in case of failures, we will exponentially backoff and retry, which will cause the rate to be lower than 1 per second due to obvious reasons.
-	r := rate.NewLimiter(rate.Every(1*time.Second), burst)
+	r := rate.NewLimiter(rate.Every(time.Duration(slackRequestRateMs)*time.Millisecond), burst)
```
Comment (ldemailly) [inline app.go:169]: use duration type directly?

File: main.go [hunk 3]
```diff
@@ -98,11 +98,13 @@ func main() {
 		metricsPort         = ":9090"
 		applicationPort     = ":8080"
 		channelOverride     string
+		slackRequestRateMs  = 1000
 	)
 
 	// Define the flags with the default values // TODO: move the ones that can change to dflag
 	flag.IntVar(&maxRetries, "maxRetries", maxRetries, "Maximum number of retries for posting a message")
 	flag.IntVar(&initialBackoffMs, "initialBackoffMs", initialBackoffMs, "Initial backoff in milliseconds for retries")
+	flag.IntVar(&slackRequestRateMs, "slackRequestRateMs", slackRequestRateMs, "Rate limit for slack requests in milliseconds")
```
Comment (ldemailly) [inline main.go:107]: duration not int

Comment (ldemailly) [reply main.go:107]: and so should have been initialBackoffMs I guess

File: app.go (hunk 1 above)
Comment (ani1311) [reply app.go]: I was thinking we could just add the fields in `App`. For example, we can have SlackURL and token in App itself. So moved it there. 

I could also add a config object like:
```
type ProcessQueueConfig struct {
    Ctx                  context.Context
    MaxRetries           int
    InitialBackoffMs     time.Duration
    Burst                int
    SlackRequestRateMs   time.Duration
}
``` 

And use it, let me know if you think that would be nicer? I was thinking 5 params is okish vs having a config object for only 4 params (assuming I don't make context a part of it.)

File: app.go (hunk 2 above)
Comment (ani1311) [reply app.go]: Done

File: main.go (hunk 3 above)
Comment (ani1311) [reply main.go]: Done

**Comment by ani1311** [comment]
Also, I saw the lint issues, will be fixing them today.

File: app.go (hunk 1 above)
Comment (ldemailly) [reply app.go]: If I count correctly from my phone I see 7 arguments in the green + the context?

but yes, I didn't see it was an App method so why not put them in the app fields?

File: app.go [hunk 4]
```diff
@@ -146,12 +146,14 @@ func (s *SlackClient) PostMessage(request SlackPostMessageRequest, url string, t
 	return nil
 }
 
-func NewApp(queueSize int, httpClient *http.Client, metrics *Metrics, channelOverride string) *App {
+func NewApp(queueSize int, httpClient *http.Client, metrics *Metrics, channelOverride, slackPostMessageUrl, slackToken string) *App {
```
Comment (ldemailly) [inline app.go:149]: ic now that's the one with many args

you could consider having caller setup all public field and having an Init() to finish the setup with no arguments

File: main.go [hunk 5]
```diff
@@ -91,18 +93,20 @@ func getSlackTokens() []string {
 func main() {
 	var (
 		maxRetries          = 2
-		initialBackoffMs    = 1000
+		initialBackoffMs    = 1000 * time.Millisecond
```
Comment (ldemailly) [inline main.go:96]: (nit) it's not Ms anymore (ditto the other)

File: main.go [hunk 6]
```diff
@@ -156,7 +160,7 @@ func main() {
 	defer serverCancel()
 
 	log.Infof("Starting main app logic")
-	go app.processQueue(ctx, maxRetries, initialBackoffMs, slackPostMessageURL, token, burst)
+	go app.processQueue(ctx, maxRetries, initialBackoffMs, burst, slackRequestRateMs)
```
Comment (ldemailly) [inline main.go:163]: unless the params change at each call (do they?) I would just have fields?

**Review by ldemailly — APPROVED** [review]

File: app.go (hunk 1 above)
Comment (ani1311) [reply app.go]: Ya the diff here has 7 args, but it is outdated. This is the updated one where I moved 2 arguments(`processQueue` and `token`) to App: https://github.com/fortio/slack-proxy/pull/42/files#diff-6c4b6ed7dc8834cef100f50dae61c30ffe7775a3f3f6f5a557517cb740c44a2dR167 

But the others I am not entirely sure about, `initialBackoffMs`, `retries` and `burst` seem like more relevant to `processQueue` than to the app itself, that's why wasn't sure if I should move them to app?

File: main.go (hunk 5 above)
Comment (ani1311) [reply main.go]: Done.

File: app.go (hunk 4 above)
Comment (ani1311) [reply app.go]: I am thinking, I can have another refactor MR where I can do these changes but for this MR we can just keep it to adding `requestRate`?

File: main.go (hunk 6 above)
Comment (ani1311) [reply main.go]: I am thinking, I can have another refactor MR where I can do these changes but for this MR we can just keep it to adding `requestRate`?

**Comment by ldemailly** [comment]
@ani1311 this broke the tests 
```
$ go test ./...
--- FAIL: TestApp_singleBurst_Success (0.00s)
    app_test.go:66: Expected processQueue finish the job in ~9 seconds, give or take. Got 4.7167e-05
--- FAIL: TestApp_MultiBurst_Success (0.00s)
    app_test.go:109: Expected processQueue finish the job in ~9 seconds, give or take. Got 2.0208e-05
--- FAIL: TestApp_TestSlackRequestRate (0.00s)
    app_test.go:152: Expected processQueue finish the job in ~5 seconds, give or take. Got 1.2083e-05
{"ts":1711044382.156533,"level":"err","r":60,"file":"main_test.go","line":20,"msg":"Testing logging - err","err":"Channel is not set and Neither attachments, blocks, nor text is set"}
{"ts":1711044382.156913,"level":"err","r":49,"file":"server.go","line":81,"msg":"Invalid request","err":"Channel is not set and Neither attachments, blocks, nor text is set"}
FAIL
FAIL	fortio.org/slack-proxy	2.204s
FAIL
```

found in #49

**Pull request merged**